File: backend/order-service/src/main/java/com/grab/order/model/Order.java
package com.grab.order.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "orders")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Order {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "customer_id", nullable = false)
    private UUID customerId;

    @Column(name = "driver_id")
    private UUID driverId;

    @Column(name = "pickup_address", nullable = false, length = 500)
    private String pickupAddress;

    @Column(name = "pickup_lat", nullable = false, precision = 10, scale = 8)
    private BigDecimal pickupLat;

    @Column(name = "pickup_lng", nullable = false, precision = 11, scale = 8)
    private BigDecimal pickupLng;

    @Column(name = "dropoff_address", nullable = false, length = 500)
    private String dropoffAddress;

    @Column(name = "dropoff_lat", nullable = false, precision = 10, scale = 8)
    private BigDecimal dropoffLat;

    @Column(name = "dropoff_lng", nullable = false, precision = 11, scale = 8)
    private BigDecimal dropoffLng;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    @Builder.Default
    private OrderStatus status = OrderStatus.PENDING;

    @Column(name = "total_price", precision = 10, scale = 2)
    private BigDecimal totalPrice;

    @Column(name = "distance_km", precision = 8, scale = 2)
    private BigDecimal distanceKm;

    @Column(length = 1000)
    private String note;

    @Column(name = "cancelled_reason", length = 500)
    private String cancelledReason;

    @OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<OrderItem> items = new ArrayList<>();

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    public enum OrderStatus {
        PENDING, ACCEPTED, PICKUP, IN_TRANSIT, DELIVERED, CANCELLED
    }
}

File: backend/order-service/internal/model/order.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type OrderStatus string

const (
	StatusPending    OrderStatus = "PENDING"
	StatusAccepted   OrderStatus = "ACCEPTED"
	StatusPickedUp   OrderStatus = "PICKED_UP"
	StatusDelivering OrderStatus = "DELIVERING"
	StatusCompleted  OrderStatus = "COMPLETED"
	StatusCancelled  OrderStatus = "CANCELLED"
)

type Order struct {
	ID               uuid.UUID   `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	CustomerID       uuid.UUID   `gorm:"type:uuid;not null;column:customer_id" json:"customer_id"`
	DriverID         *uuid.UUID  `gorm:"type:uuid;column:driver_id" json:"driver_id,omitempty"`
	Status           OrderStatus `gorm:"size:20;not null;default:'PENDING'" json:"status"`
	PickupAddress    string      `gorm:"column:pickup_address;not null" json:"pickup_address"`
	PickupLat        float64     `gorm:"column:pickup_lat" json:"pickup_lat"`
	PickupLng        float64     `gorm:"column:pickup_lng" json:"pickup_lng"`
	DropoffAddress   string      `gorm:"column:dropoff_address;not null" json:"dropoff_address"`
	DropoffLat       float64     `gorm:"column:dropoff_lat" json:"dropoff_lat"`
	DropoffLng       float64     `gorm:"column:dropoff_lng" json:"dropoff_lng"`
	Note             string      `gorm:"size:500" json:"note,omitempty"`
	TotalPrice       float64     `gorm:"column:total_price" json:"total_price"`
	CreatedAt        time.Time   `json:"created_at"`
	UpdatedAt        time.Time   `json:"updated_at"`
}

func (Order) TableName() string { return "orders" }
